### ISSUE TITLE
Segmentation fault (core dumped) while using bsrate 

### ISSUE THREAD

**Issue by Akshaygala** [smithlabcode/methpipe#199]
I am trying to use bsrate function and am getting segmentation fault error. Here is the glimpse of my code:
"
INbam="${!SGE_TASK_ID}"
samtools view -H "${INbam/.bam/}".sam > "${INbam/.bam/}".headers.txt                                                    && \
awk -F "\t" '{if ($3=="MT") print}' "${INbam/.bam/}".sam > "${INbam/.bam/}".reads.sam                          && \
cat "${INbam/.bam/}".headers.txt "${INbam/.bam/}".reads.sam > "${INbam/.bam/}".chrMT.sam               && \
rm "${INbam/.bam/}".headers.txt "${INbam/.bam/}".reads.sam                                                                      && \

bsrate -c "${INbam/.bam/}".chrMT.sam -o "${INbam/.bam/}".bsrate "${INbam/.bam/}".sam
"

I am providing the bam files using an array job. My chrMT.sam files have headers text followed by the MT reads. I have attached my input file for bsrate. Any help will be appreciated. 
[S9_.bsbolt-grch38.sorted.deduped.chrMT.txt](https://github.com/smithlabcode/methpipe/files/7937688/S9_.bsbolt-grch38.sorted.deduped.chrMT.txt)

Thanks.

Akshay

**Comment by guilhermesena1** [comment]
Hi,

Can you confirm which version of methpipe you are using?

**Comment by Akshaygala** [comment]
I am using Methpipe-5.0.0

**Comment by guilhermesena1** [comment]
oh I see in your command that you seem to be passing a SAM file as the reference? The `-c` option should be the same FASTA file you used to map your reads or, alternatively, just chrM since you are extracting only the reads that map to that chrom (either should work)

**Comment by Akshaygala** [comment]
Yes. For the -c option, I am passing the sam file which contains chromosome M reads including header. It is the same file that I have attached in my first post.

**Comment by guilhermesena1** [comment]
right, so I think the order of parameters is switched in your command. The attached file should be the last argument (which in the command you posted is given by `"${INbam/.bam/}".sam`). I think this is your whole reads file? If it is, then you don't need it to estimate bsrate based on chrM.

There are two ways to run bsrate, (1) one is based on the Cs that map to non-CpGs in the whole reference genome, and (2) one based on all Cs that map to mitochondrial DNA (for example, if your organism is expected to exhibit non-CpG methylation).

Suppose you get a file `out.sam` mapped to `genome.fa`, and that you extracted chrM files making file `out_mt.sam`

To obtain (1) you run
```
bsrate -c genome.fa -o out.bsrate out.sam
```

To obtain (2) you run
```
bsrate -N-c genome.fa -o out_mt.bsrate out_mt.sam
```
the -N parameter above tells bsrate to also consider converted CpGs as failed conversion, which you will want to do in mitochondrial DNA.

I hope that helps

**Comment by Akshaygala** [comment]
@guilhermesena1 Ohh Snap.. I misinterpreted the syntax. Thank you for the correction. It works (y)

Akshay

**Comment by andrewdavidsmith** [comment]
This shouldn't segfault anyway, so I'm transferring this issue.
